### PR TITLE
Mel loss/bill

### PR DESCRIPTION
**Pull request by billyang98**
added new mel loss to incorporate pre conv postnet mels predictions as per tacotron and the transformer tts repo 

"We minimize the summed mean squared error (MSE) from before
and after the post-net to aid convergence"

https://github.com/soobinseo/Transformer-TTS/blob/7a1f23baa8cc703f63cc2f11405f6898e3217865/train_transformer.py#L49-L50

Also merged discriminator changes with all new changes

### CODE REVIEW

File: src/network.py [hunk 1]
```diff
@@ -94,44 +94,53 @@ def __init__(self, text_m, speech_m, discriminator=None, teacher=None):
...
         text_pred, text_pred_len = self.text_m.infer_sequence(s_e_o, s_mask)
         cm_t_e_o, cm_t_masks = self.text_m.encode(text_pred, text_pred_len)
-        pred, stop_pred = self.speech_m.decode_sequence(mel, mel_len, cm_t_e_o, 
+        pre_pred, post_pred, stop_pred = self.speech_m.decode_sequence(mel, mel_len, cm_t_e_o,
             cm_t_masks, teacher_ratio=self.teacher.get_val())
-        return pred, stop_pred
-    
-    def tts(self, text, text_len, mel, mel_len, infer=False):
+        if ret_enc_hid:
+            return pre_pred, post_pred, stop_pred, s_e_o[0], cm_t_e_o[0]
```
Comment (Lucaskabela) [inline src/network.py]: I don't think s_e_o[0] will correctly fetch the hidden state.  encoder outputs should be [batch_size x seq_len x hidden_dim]!

File: src/train.py [hunk 2]
```diff
@@ -104,55 +104,153 @@ def speech_loss(gold_mel, stop_label, pred_mel, mel_len, stop_pred):
...
+    target = discriminator_target(d_out, target_type)
+    d_loss = discriminator_loss(d_out, target)
+    return d_loss
+
+def discriminator_step(model, batch):
+    x, _ = process_batch(batch)
+    text, mel, text_len, mel_len = x
+
+    # text
+    (t_hid, _), _ = model.text_m.encode(text, text_len)
```
Comment (Lucaskabela) [inline src/train.py]: This logic won't work for transfomers :(

File: src/train.py [hunk 3]
```diff
@@ -163,39 +261,86 @@ def optimizer_step(model, optimizer, args):
...
 
     # Log losses
-    losses['s_cm'].append(s_cm_loss.detach().item())
-    losses['t_cm'].append(t_cm_loss.detach().item())
+    losses['s_cm'].append(s_cm_loss.detach().cpu().item())
+    losses['t_cm'].append(t_cm_loss.detach().cpu().item())
+    if args.use_discriminator:
+        losses['s_cm_d'].append(s_d_loss.detach().cpu().item())
+        losses['t_cm_d'].append(t_d_loss.detach().cpu().item())
+        losses['cm_s_cm_d'].append(cm_s_d_loss.detach().cpu().item())
```
Comment (Lucaskabela) [inline src/train.py]: In keeping with there should be no gradients for the first step of the cross models (so text encode speech decode for text cross model and similarly for speech cross model) I think we shouldn't have this many discriminator losses...  Should we cut the loss from the first encoder?

File: src/train.py [hunk 4]
```diff
@@ -163,39 +261,86 @@ def optimizer_step(model, optimizer, args):
...
+    t_d_loss, s_d_loss = discriminator_step(model, batch)
+
+    loss = (t_d_loss + s_d_loss) / accum_steps
+    loss.backward()
+
+    losses['t_d'].append(t_d_loss.detach().cpu().item())
+    losses['s_d'].append(s_d_loss.detach().cpu().item())
+    return loss
+
+def freeze_model_parameters(model):
```
Comment (Lucaskabela) [inline src/train.py:470]: Can we use .eval() and .train() to achieve this freeze/unfreeze

**Review by Lucaskabela — CHANGES_REQUESTED** [review]

File: src/network.py (hunk 1 above)
Comment (billyang98) [reply src/network.py]: encoder outputs is a tuple   


 def encode(self, text, text_lens, noise_in=False):
        """
        Returns:
            - output: a tensor of shape [batch_size x seq_len x latent_dim], which
                is the projected hidden state from each timestep in the sequence
            - h_t: the (h_n, c_n) pair from the last timestep of the network.
                Dimension varies depending on num_layers, but should
                be proportional to  [num_layers, batch, hidden_size]
            - pad_mask: a tensor of shape [batch_size x seq_len] with 1s for padded
                indices and 0 for everything else
        """
        embedded_phonemes, pad_mask = self.preprocess(text, text_lens, noise_in)
        enc_output, hidden_state = self.encoder(embedded_phonemes)
        return (hidden_state, enc_output), pad_mask

File: src/train.py (hunk 4 above)
Comment (billyang98) [reply src/train.py:470]: No, unfortunately torch doesn't like calling eval() on the encoders and decoders and then trying to do an optimizer step.

File: src/network.py (hunk 1 above)
Comment (Lucaskabela) [reply src/network.py]: Ah, I was still thinking of transformers where this will be  [batch_size x seq_len x hidden_dim].   note for when we add discriminator support for trasformers then!

File: src/train.py (hunk 4 above)
Comment (Lucaskabela) [reply src/train.py:470]: Oh, I see - this is a result of all parameters using the same optimizer. Well that is unfortunate 😞

File: src/train.py [hunk 5]
```diff
@@ -170,39 +277,84 @@ def optimizer_step(model, optimizer, args):
...
-def train_cm_step(losses, model, batch, accum_steps):
+def train_cm_step(losses, model, batch, accum_steps, use_discriminator):
+    # NOTE: do not use cross_model here bc need to take optimizer step inbetween
     batch = process_batch(batch)
-    t_cm_loss, s_cm_loss = crossmodel_step(model, batch)
-    loss = (s_cm_loss + t_cm_loss) / accum_steps
+
+    if use_discriminator:
+        t_cm_loss, s_cm_loss, cm_t_d_loss, cm_s_d_loss = crossmodel_step(model, batch, use_discriminator)
+        loss = s_cm_loss + t_cm_loss + s_d_loss + cm_t_d_loss + t_d_loss + cm_s_d_loss
```
Comment (Lucaskabela) [inline src/train.py]: Still 6 losses here - will cause an error (?)

File: src/train.py [hunk 6]
```diff
@@ -104,62 +104,169 @@ def masked_mse(gold_mel, pred_mel, mel_mask):
...
+    target = discriminator_target(d_out, target_type)
+    d_loss = discriminator_loss(d_out, target)
+    return d_loss
+
+def discriminator_step(model, batch):
+    x, _ = process_batch(batch)
+    text, mel, text_len, mel_len = x
+
+    # text
+    t_enc_out, _ = model.text_m.encode(text, text_len)
```
Comment (Lucaskabela) [inline src/train.py]: Can we call these in a context manager, depending on who is getting updated?  So if it is discriminator's turn:
`
if train_d:
  with torch.no_grad():
       t_enc_out, _ = model.text_m.encode(text, text_len)
else:
       t_enc_out, _ = model.text_m.encode(text, text_len)
`
Helps to make sure we don't give gradients to our encoder, and makes the code run faster!

**Review by Lucaskabela — CHANGES_REQUESTED** [review]
Looks great! - one potential error spotted, and one nit/request

File: src/train.py [hunk 7]
```diff
@@ -104,62 +104,169 @@ def masked_mse(gold_mel, pred_mel, mel_mask):
...
+    # quick check to determine between rnn and transformer
+    # eventually should be built into the RNN and Transformer Encoder classes
+    if len(t_enc_out == 2):
```
Comment (Lucaskabela) [inline src/train.py]: len(t_enc_out) == 2

Comment (Lucaskabela) [reply src/train.py]: Same issue for s_enc_out a few lines later

**Comment by Lucaskabela** [comment]
As another note, I got this error trying to train the transformer model:
  File "/content/UNAST/src/network.py", line 98, in text_ae
    return self.text_m.forward(text, text_len, noise_in=True, teacher_ratio=1, ret_enc_hid=ret_enc_hid)

**Comment by billyang98** [comment]
Nice catches Lucas! I haven't been able to test this fully yet, but the core idea should be good.

I might not get around to testing tomorrow, but take a look or test yourself in the meantime.
Go ahead and push directly to the branch too if you want any changes.

File: src/train.py [hunk 8]
```diff
@@ -104,62 +104,175 @@ def masked_mse(gold_mel, pred_mel, mel_mask):
...
     t_cm_loss = text_loss(gold_char.to(DEVICE), text_pred)
+
+    if use_dis_loss:
+        return t_cm_loss, s_cm_loss, cm_t_d_loss, cm_s_d_loss
     return t_cm_loss, s_cm_loss
 
+def discriminator_hidden_to_loss(model, hid, target_type, freeze_discriminator= False):
+    d_in = hid[-1]
+    if freeze_discriminator:
+        with torch.no_grad():
```
Comment (billyang98) [inline src/train.py:332]: I'm trying this approach to "freeze" the discriminator during training. It should be faster, but I'm not 100% that this is doing exactly what we want.

**Review by Lucaskabela — APPROVED** [review]
